src/transformers/models/qwen2_vl/modeling_qwen2_vl.py

```diff
@@ -484,6 +488,165 @@ def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
+    # Replace inf values with zeros in attention weights to prevent NaN propagation
+    if query.dtype == torch.float16:
+        attn_weights = torch.where(torch.isinf(attn_weights), torch.zeros_like(attn_weights), attn_weights)
+
+    # upcast attention to fp32
+    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query.dtype)
+    attn_weights = nn.functional.dropout(attn_weights, p=config.attention_dropout, training=config.training)
+    attn_output = torch.matmul(attn_weights, value_states)
+
+    attn_output = attn_output.transpose(1, 2).contiguous()
+    return attn_output, attn_weights
+
+
+def flash_attention_forward(
+    config: Qwen2VLConfig,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    mask: Optional[torch.Tensor],
+    target_dtype: torch.dtype = torch.float16,
+    **_kwargs,
+) -> Tuple[torch.Tensor, None]:
+    key_states = repeat_kv(key, config.num_key_value_groups)
+    value_states = repeat_kv(value, config.num_key_value_groups)
+    dropout_rate = 0.0 if not config.training else config.attention_dropout
+
+    # In PEFT, usually we cast the layer norms in float32 for training stability reasons
+    # therefore the input hidden states gets silently casted in float32. Hence, we need
+    # cast them back in float16 just to be sure everything works as expected.
+    input_dtype = query.dtype
+    if input_dtype != target_dtype:
+        query = query.to(target_dtype)
+        key_states = key_states.to(target_dtype)
+        value_states = value_states.to(target_dtype)
+
+    # Reashape to the expected shape for Flash Attention
+    query_states = query.transpose(1, 2)
+    key_states = key_states.transpose(1, 2)
+    value_states = value_states.transpose(1, 2)
+
+    position_ids = _kwargs["position_ids"]
+
+    attn_output = _flash_attention_forward(
+        query_states,
+        key_states,
+        value_states,
+        mask,
+        query_states.shape[1],
+        position_ids=position_ids,
+        dropout=dropout_rate,
+        sliding_window=config.sliding_window,
+        is_causal=config.is_causal,
+        use_top_left_mask=config._flash_attn_uses_top_left_mask,
+    )
+
+    attn_output = attn_output.to(input_dtype)
+    return attn_output, None
+
+
+def flex_attention_forward(
+    config: Qwen2VLConfig,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    mask: Optional[torch.Tensor],
+    **_kwargs,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    key_states = repeat_kv(key, config.num_key_value_groups)
+    value_states = repeat_kv(value, config.num_key_value_groups)
+
+    causal_mask = mask
+    if causal_mask is not None:
+        causal_mask = causal_mask[:, :, :, : key_states.shape[-2]]
+
+    def causal_mod(score, b, h, q_idx, kv_idx):
+        if causal_mask is not None:
+            score += causal_mask[b][0][q_idx][kv_idx]
+        return score
+
+    attn_output, attn_weights = flex_attention(
```

Comment:
I'm having a hard time getting `attn_weights` to match eager. Can I merge the refactor for now, and add flex attention as a follow up?